app/server/models/transfer.py

```diff
@@ -0,0 +1,611 @@
+                return approval
+        return None
+
+    @hybrid_property
+    def total_sent(self):
+        return int(
+            db.session.query(func.sum(CreditTransfer.transfer_amount).label('total')).execution_options(show_all=True)
+            .filter(CreditTransfer.transfer_status == TransferStatusEnum.COMPLETE)
+            .filter(CreditTransfer.sender_transfer_account_id == self.id).first().total or 0
+        )
+
+    @hybrid_property
+    def total_received(self):
+        return int(
+            db.session.query(func.sum(CreditTransfer.transfer_amount).label('total')).execution_options(show_all=True)
+            .filter(CreditTransfer.transfer_status == TransferStatusEnum.COMPLETE)
+            .filter(CreditTransfer.recipient_transfer_account_id == self.id).first().total or 0
+        )
+
+    # @hybrid_property
+    # def balance(self):
+    #     return self.total_received - self.total_sent
+
+    @hybrid_property
+    def primary_user(self):
+        users = User.query.execution_options(show_all=True) \
+            .filter(User.transfer_accounts.any(TransferAccount.id.in_([self.id]))).all()
+        if len(users) == 0:
+            # This only happens when we've unbound a user from a transfer account by manually editing the db
+            return None
+
+        return sorted(users, key=lambda user: user.created)[0]
+
+    @hybrid_property
+    def primary_user_id(self):
+        return self.primary_user.id
+
+    @hybrid_property
+    def master_wallet_approval_status(self):
+
+        if not current_app.config['USING_EXTERNAL_ERC20']:
+            return 'NOT_REQUIRED'
+
+        if not self.blockchain_address.encoded_private_key:
+            return 'NOT_REQUIRED'
+
+        base_query = (
+            BlockchainTransaction.query
+                .filter(BlockchainTransaction.transaction_type == 'master wallet approval')
+                .filter(BlockchainTransaction.credit_transfer.has(recipient_transfer_account_id=self.id))
+        )
+
+        successful_transactions = base_query.filter(BlockchainTransaction.status == 'SUCCESS').all()
+
+        if len(successful_transactions) > 0:
+            return 'APPROVED'
+
+        requested_transactions = base_query.filter(BlockchainTransaction.status == 'PENDING').all()
+
+        if len(requested_transactions) > 0:
+            return 'REQUESTED'
+
+        failed_transactions = base_query.filter(BlockchainTransaction.status == 'FAILED').all()
+
+        if len(failed_transactions) > 0:
+            return 'FAILED'
+
+        return 'NO_REQUEST'
+
+    def approve(self):
+
+        if not self.is_approved:
+            self.is_approved = True
+
+            if self.is_beneficiary:
+                disbursement = self.make_initial_disbursement()
+                return disbursement
+
+    def make_initial_disbursement(self, initial_balance=None):
+        #TODO(refactor): circular dependency
+        from server.utils.credit_transfers import make_disbursement_transfer
```

Comment:
what's the purpose of utils.user and utils.credit_transfer? what distinguishes helper functions in those vs functions on the models for them respectively? those were points of contention for circular dependencies. see the other "TODO(refactor)"s. is this the correct way to fix them?

Reply:
actually, i'm kinda confused now when an import causes an error for dependency vs is okay with it. models.py before was working just fine but there must have also been loops with the utils right?

Reply:
To add to my comments at the bottom of the thread:

> What distinguishes helper functions in those vs static functions on the models? 

In short, I built them out there to stop the models file from getting way huge, but your refactor is a better solution.
I don't _love_ doing database actions (such as db.session.add() )from inside database models themselves, but tbh I'm not convinced there's any good logic behind that; the only db action that I know should definitely be extracted out is session commit.
It may actually be reasonable to shift those into the models themselves, now especially if they reduce import errors.

> adding an import inside the function instead of at the top. that's what i read somewhere but will this cause an error when the function itself is run?

Yeah in-function imports are pretty much a last resort. My Pycharm Linter actually yells at me if i try and do them lol.